arrow-fx-coroutines/README.MD

```diff
@@ -1,561 +1,561 @@
+
+The `DefaultConnection` keeps track of all `CancelToken` registered for a given `SuspendConnection`.
+It does so by keeping all `CancelToken` in a FIFO stack, and running them in order when cancelled.
+Since `SuspendConnection` is a FIFO stack, we can register a token before running a cancellable operation,
+and pop the token after the cancellable operation has finished so
+we don't run a `CancelToken` when it's not necessary anymore since that could lead to (undefined) weird behavior. This is useful for example in `parMap2` where we want to launch two parallel operations, and push their cancellation system onto to the original `SuspendConnection` and `pop` it when both finish.
+
+### Cancellation
+
+So what does it actually mean to cancel? Currently in Arrow 0.11.0, cancellation means that the code stops running (or in other words suspends forever and `never` returns).
+This is implemented as a `cancelBoundary`, and with all the understanding of the low-level Kotlin machinery we have now can be implemented as follows.
+
+```kotlin
+suspend fun cancelBoundary(): Unit =
+  suspendCoroutineUninterceptedOrReturn { cont ->
+    if (cont.context[SuspendConnection]?.isCancelled() == true) COROUTINE_SUSPENDED
+    else Unit
   }
-}
 ```
 
-## Parallel Ops
+So a `cancelBoundary` checks if the `SuspendConnection#isCancelled` and if so it suspends forever and never invokes the callback, effectively this means it stops running the code.
+In the case the `SuspendConnection` isnt' cancelled it immediately returns `Unit`. Here we use `suspendCoroutineUninterceptedOrReturn` since we can return immediately when no cancellation signal was triggered, and we don't want to intercept or schedule but we have to check the `Continuation#context` to be able to access the `SuspendConnection`.
 
+So this is how from any user or library code we can add checks for cancellation that automatically exit when a cancellation signal was trigerred.
 
-### ForkConnected
+When `isCancelled` returns `true`, this means that all the registered `CancelToken`'s inside `SuspendConnection` have run, or are running.
 
-`forkConnected` launches a `suspend () -> A` function on a given `CoroutineContext` inside a `Fiber`, the lifecycle of this `Fiber` is connected to its parent.
-If the parent cancels that this operation will also be cancelled.
+NOTE:
+In Kotlin 1.4, CancellationException was added to the Kotlin Std as an ExperimentalApi. The documentation states that a cancellable coroutine should throw it to signal that it has been canccelled. In a future version of Arrow Fx Coroutines we can follow this pattern by instead of suspend upon cancellation like we did in Arrow Fx IO, and as is known from Cats Effects IO, we can `throw CancellationException`. This would also only happen when we're running on a cancellable `SuspendConnection`, and we'd still be able to control cancellation in the same way as we know now are you might be familiar with from Cats Effects IO or Arrow Fx IO.
```

Comment:
```suggestion
In Kotlin 1.4, CancellationException was added to the Kotlin Std as an ExperimentalApi. The documentation states that a cancellable coroutine should throw it to signal that it has been canccelled. In a future version of Arrow Fx Coroutines we can follow this pattern by instead of suspend upon cancellation like we did in Arrow Fx IO, and as is known from Cats Effects IO, we can `throw CancellationException`. This would also only happen when we're running on a cancellable `SuspendConnection`, and we'd still be able to control cancellation in the same way as you might be familiar with from Cats Effects IO or Arrow Fx IO.
```